### PR TITLE
Uvl 85 inbound data handler

### PR DESCRIPTION
**Pull request by mattrobertsky**

### CODE REVIEW

File: app/uk/gov/hmrc/vatregisteredcompanies/models/PayloadSubmissionResponse.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+package uk.gov.hmrc.vatregisteredcompanies.models
+
+import java.time.{LocalDateTime, ZoneId}
+
+import play.api.libs.json.{Json, OFormat}
+
+case class PayloadSubmissionResponse(
+  outcome: String,
+  code: Option[String],
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/models/PayloadSubmissionResponse.scala]: consider enums

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import uk.gov.hmrc.vatregisteredcompanies.models
+import uk.gov.hmrc.vatregisteredcompanies.models.{LookupResponse, Payload, VatNumber, VatRegisteredCompany}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import cats.implicits._
+
+final case class Wrapper(
+  vatNumber: VatNumber,
+  company: VatRegisteredCompany
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: doesn't `VatRegisteredCompany` contain the `VatNumber` anyway?

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import reactivemongo.play.json.ImplicitBSONHandlers._
+import uk.gov.hmrc.mongo.ReactiveRepository
+import uk.gov.hmrc.vatregisteredcompanies.models
+import uk.gov.hmrc.vatregisteredcompanies.models.{LookupResponse, Payload, VatNumber, VatRegisteredCompany}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import cats.implicits._
+
+final case class Wrapper(
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: What is this for?

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    deletes.map{ vatNumber =>
+      remove("vatNumber" -> vatNumber)
+    }.map(_ => Future(())).head
+
+  private def wrap(payload: Payload): List[Wrapper] =
+    payload.createsAndUpdates.map { company =>
+      Wrapper(company.vatNumber, company)
+    }
+
+  def process(payload: Payload): Future[(Unit, Unit)] = {
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: Consider - 
```
insert(wrap(payload)) >> delete(payload.deletes)
```

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  }.map(_ => ())
+
+  private def insert(entries: List[Wrapper])=
+    entries.map(wrapper => insertOrUpdate(wrapper)).head
+
+  private def delete(deletes: List[VatNumber]) =
+    deletes.map{ vatNumber =>
+      remove("vatNumber" -> vatNumber)
+    }.map(_ => Future(())).head
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: Consider something like - 
```
private def delete(deletes: List[VatNumber]): Future[Unit] = 
  deletes.traverse_{ vatNumber =>remove("vatNumber" -> vatNumber)}
```

File: app/uk/gov/hmrc/vatregisteredcompanies/services/PersistenceService.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.inject.{Inject, Singleton}
+import play.api.mvc.Result
+import uk.gov.hmrc.vatregisteredcompanies.models.{LookupResponse, Payload, VatNumber}
+import uk.gov.hmrc.vatregisteredcompanies.repositories.VatRegisteredCompaniesRepository
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class PersistenceService @Inject()(repository: VatRegisteredCompaniesRepository)(implicit executionContext: ExecutionContext) {
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/services/PersistenceService.scala]: Maybe an abstract trait that is not coupled to Mongo would be good here?

File: app/uk/gov/hmrc/vatregisteredcompanies/test/TestController.scala [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+
+import scala.concurrent.Future
+
+class TestController @Inject()(
+  connector: TestConnector
+) extends FrontendController {
+
+  def triggerDataImport: Action[AnyContent] = Action.async { implicit request =>
+    connector.trigger("trigger-mdg-data-post") flatMap (_ =>
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatregisteredcompanies/test/TestController.scala]: `a >> b` is the same as `a.flatMap{ _ => b}`. It's a bit clearer imho but might confuse people who have not come across it before.

File: conf/app.routes [hunk 8]
```diff
@@ -1,3 +1,8 @@
 # microservice specific routes
 
-GET        /hello-world             uk.gov.hmrc.vatregisteredcompanies.controllers.MicroserviceHelloWorld.hello
+GET        /hello-world               uk.gov.hmrc.vatregisteredcompanies.controllers.MicroserviceHelloWorld.hello
```
Comment (hmrclt) [inline conf/app.routes]: Do you still have a hello world controller? Do you need it?

File: conf/application.conf [hunk 9]
```diff
@@ -157,6 +170,11 @@ microservice {
             port=8500
         }
 
+        mdg { # used for mdg and stub
+            inboundData.token = "ab9e219d-0d9d-4a1d-90e5-f2a5c287668c"
```
Comment (hmrclt) [inline conf/application.conf]: please don't put this in the code

File: app/uk/gov/hmrc/vatregisteredcompanies/controllers/ExtraActions.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import play.api.http.HeaderNames
+import play.api.mvc.Results.Unauthorized
+import play.api.mvc.{ActionBuilder, ActionFilter, Request, Result}
+import uk.gov.hmrc.play.config.ServicesConfig
+
+import scala.concurrent.Future
+
+trait ExtraActions extends ServicesConfig {
+
+  val InboundDataAction: ActionBuilder[Request] = AuthorisedFilterAction
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/controllers/ExtraActions.scala]: Probably as I would anticipate having multiple FilterActions. Then I can extend the ExtraActions trait and cherry pick which FilterAction, or combination thereof, that I use.

File: app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      persistence.lookup(target).map { x =>
+        Ok(Json.toJson(x.getOrElse(LookupResponse(None))))
+      }
+    }
+
+  def lookupVerified(target: VatNumber, requester: VatNumber): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val targetLookup = persistence.lookup(target)
+      val requesterLookup = persistence.lookup(requester)
+      val futureResult: Future[Option[LookupResponse]] = for {
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala]: Will those two persistence calls run in parallel or sequentially?

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala (hunk 2 above)
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: Yes, but I'm using the vat number as the index on the collection.

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala (hunk 3 above)
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: There's a .some on line 94.

File: conf/application.conf (hunk 9 above)
Comment (mattrobertsky) [reply conf/application.conf]: commented as discussed

File: app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala (hunk 5 above)
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/repositories/VatRegisteredCompaniesRepository.scala]: Done

File: app/uk/gov/hmrc/vatregisteredcompanies/models/PayloadSubmissionResponse.scala (hunk 1 above)
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/models/PayloadSubmissionResponse.scala]: As discussed parking this for now.

File: app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        a <- targetLookup
+        b <- requesterLookup
+      } yield a.map(x => x.copy(
+        requester = b.fold(Option.empty[VatNumber])(_ => Some(requester)),
+        consultationNumber = b.fold(Option.empty[ConsultationNumber])(_ => Some(ConsultationNumber.generate))
+      ))
+
+      futureResult.map {x =>
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala]: done - I think

File: app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import uk.gov.hmrc.play.bootstrap.controller.BaseController
+import uk.gov.hmrc.vatregisteredcompanies.models.{ConsultationNumber, LookupResponse, VatNumber}
+import uk.gov.hmrc.vatregisteredcompanies.services.PersistenceService
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class VatRegCoLookupController @Inject()(persistence: PersistenceService, auditConnector: AuditConnector)(implicit executionContext: ExecutionContext) extends BaseController {
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/controllers/VatRegCoLookupController.scala]: done

File: app/uk/gov/hmrc/vatregisteredcompanies/controllers/InboundDataController.scala [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class InboundDataController @Inject()(persistence: PersistenceService)(implicit executionContext: ExecutionContext, conf: Configuration, environment: Environment)
+  extends BaseController with ExtraActions {
+
+  def handle: Action[JsValue] =
+    InboundDataAction.async(parse.json) { implicit request =>
+      withJsonBody[Payload] {
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/vatregisteredcompanies/controllers/InboundDataController.scala]: done

**Pull request merged**